### DISCUSSION TITLE
bTrack cell labeling and mitosis detection problems

### DISCUSSION THREAD

**Discussion by emthilak** [quantumjot/btrack#339] [Q&A]
Hi,

I've been using bTrack to track cells over long time courses and visualize them using napari, but I've been running into two (related) issues.  They are as follows:

1. bTrack labels single pixels as cells.  This appears to be a consequence of discontinuous cell masks at the periphery of cells (masks generated by CellPose).  I want to be able to discount them when using the btrack.utils.segmentation_to_objects() function (perhaps imposing a minimum area requirement for cell labeling?), but this doesn't appear to be an option.  How might I be able to remedy this issue without altering the cell masks I've generated?

2. bTrack fails to keep track of mitosis events for cells of expected size, however fluctuations in the falsely detected single pixel "cells" are detected as mitosis events.  For cells of expected size, the parent cell is tracked and after division, one of the daughter cells keeps the same track as the parent while a new track is generated for the other daughter cell.  As a consequence, lineage is indeterminate and absent in the data for relevant tracks.  Conversely for the false cells, if a false cell is detected a frame or two before multiple false cells are detected in the nearby area, a lineage is established by bTrack.  Why is bTrack generating lineages for these false cells while the expected cells that undergo mitosis do not get any?  And what can I do to fix the situation?

I've attached my configuration file, but I'm happy to provide any additional materials from my work.  Thank you!

[cell_config -test 3.txt](https://github.com/quantumjot/btrack/files/11941506/cell_config.-test.3.txt)

**Comment by quantumjot** [comment]
Hi @emthilak,

[I converted this to a discussion so that others may be able to use it in the future]

Thanks for using btrack! It can sometimes be difficult to advise without seeing the data, but I can give you a few pointers that may help.

1. You can filter the objects in the following way:

```python
objects = btrack.utils.segmentation_to_objects(
    segmentation, 
    properties=("area", ),
)

# this will create a new list of objects that have segmented areas greater than 100
filtered_objects = [obj for obj in objects if obj.properties["area"]>100]

with btrack.BayesianTracker() as tracker:
    ...

    # append the filtered objects to be tracked
    tracker.append(filtered_objects)
```

2. The filtering may help a lot in making the lineaging better. The other thing you can try is to set the `prob_not_assign` to a higher value:

```json
"prob_not_assign": 0.1
```

This will stop the tracker connecting detections too easily, so it may improve the end result.

Hope that helps!

**Reply by emthilak** [reply]
Hi,

Thanks for the quick response.  I tried filtering as you suggested and it worked well with my data, thank you!  I decided to keep `prob_not_assign` as 0.001 as higher values are not optimal for my data.

As for the second issue, I'll provide some visual data from napari here for your convenience (thank you for bearing with me).  Here is one example of many instances of this issue:

Cell 5 (ID 5) is a cell that undergoes mitosis during the course of the experiment, splitting into IDs 5 and 84, as follows:
![image](https://github.com/quantumjot/btrack/assets/98061305/614f8064-6831-4311-9120-15196235b8c0)

![image](https://github.com/quantumjot/btrack/assets/98061305/c00aef9e-dab6-440b-97e3-863a981abcdf)

![image](https://github.com/quantumjot/btrack/assets/98061305/ba59a0d6-6439-442e-88b9-f8a69c6fe232)

![image](https://github.com/quantumjot/btrack/assets/98061305/590aa258-bc9a-47a2-9307-951a978f1f0d)

I have seen other implementations of bTrack instead have such an event starting with one ID for the parent cell and two new IDs for the daughter cells.  However, as is apparent in the images, one of the daughter cells maintains the same track as the parent cell, and thus a mitosis event is not detected (and the graph variable which should contain parent-child relationships remains an empty dictionary).  What might be going wrong here?  I'll add a relevant portion of my code below as well if that may help.  Thank you!

Code:
```
import btrack
path = r'inputdir\cell_config -test-3.json'  

#running bTrack and selecting the cells that have complete tracks and no loss of segmentation.  
seg_masks=np.zeros((len(seg_select),2048,2048)) #seg_select is a list containing file paths to cell mask arrays for the experiment
for idx, image in enumerate(seg_select):
    seg_masks[idx,:,:]= np.loadtxt(image)
    
seg_masks=seg_masks.astype(int)  #convert float values in the array to integers

#running bTrack  
#localising the objects. 
obj_from_arr = btrack.utils.segmentation_to_objects(seg_masks,properties = ('area', 'major_axis_length','minor_axis_length','perimeter'),assign_class_ID=True)

#filtering false detections
filtered_objs = [obj for obj in obj_from_arr if obj.properties["area"]>20]


#%% running btrack with the objects. 
with btrack.BayesianTracker() as tracker:
        
    # configure the tracker using a config file
    tracker.configure_from_file(path)
    tracker.max_search_radius = 100

    # append the objects to be tracked
    tracker.append(filtered_objs)

    # set the volume
    tracker.volume=((0, 2048), (0, 2048), (-1e5, 1e5))

    # track them (in interactive mode)
    tracker.track_interactive(step_size=100)

    # generate hypotheses and run the global optimizer
    tracker.optimize()

    # get the tracks in a format for napari visualization
    data, properties, graph = tracker.to_napari(ndim=2)
        
#%% Visual testing

import napari

viewer = napari.Viewer()
viewer.add_image(raw_img_gfp) #raw_img_gfp contains images of the cells for every timepoint in the experiment
viewer.add_labels(seg_masks)
viewer.add_tracks(data, properties=properties, graph=graph)
```
